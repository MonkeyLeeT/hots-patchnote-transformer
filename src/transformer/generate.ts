import _ from 'lodash';

import { findNearestColor } from './color';
import { Attribute, Node } from './const';

function parseStyle(node: Node): Record<string, string> {
  if (node.kind === 'text') {
    return {};
  }
  const styleAttr = node.attrs.find((v: Attribute) => v.name === 'style');
  if (!styleAttr) {
    return {};
  }
  const styles = styleAttr.value
    .split(';')
    .map(_.trim)
    .filter(Boolean)
    .map((s: string) => s.split(':'))
    .map(([a, b]: string[]) => [_.trim(a), _.trim(b)])
    .filter((v: string[]) => v.length === 2);

  return _.fromPairs(styles);
}

function findAttr(node: Node, field: string): Attribute | undefined {
  if (node.kind === 'text') {
    return undefined;
  }

  return node.attrs.find((attr: Attribute) => attr.name === field);
}

function getNumAttr(node: Node, attr: string) {
  if (node.kind === 'text') {
    return null;
  }
  const attrObj = node.attrs.find((v: Attribute) => v.name === attr);
  if (!attrObj) {
    return null;
  }

  return attrObj.value;
}

function trimNewlines(str: string): [string, number, number] {
  const trimmedLeft = _.trimStart(str, '\n');
  const left = str.length - trimmedLeft.length;
  const trimmedRight = _.trimEnd(trimmedLeft, '\n');
  const right = trimmedLeft.length - trimmedRight.length;

  return [trimmedRight, left, right];
}

function ensureNewlines(nodeText: string, startNum: number, endNum: number = 0) {
  const [trimmed, left, right] = trimNewlines(nodeText);

  return _.repeat('\n', Math.max(startNum, left)) + trimmed + _.repeat('\n', Math.max(endNum, right));
}

// tslint:disable-next-line:max-func-body-length cyclomatic-complexity
export function generateBbsCode(node: Node, context: Record<string, unknown> = {}) {
  if (node.kind === 'text') {
    return node.text;
  }
  // <picture>: pick only the first child
  if (node.tag === 'picture') {
    node.children = node.children.slice(0, 1);
  }
  // Node types to skip
  if (_.includes(['style', 'script', '#comment'], node.tag)) {
    return '';
  }
  const style = parseStyle(node);
  const childrenRaw = generateBbsNodeList(node.children);
  let children = childrenRaw;
  if (style.color) {
    const ngaColor = findNearestColor(style.color);
    if (ngaColor) {
      children = `[color=${ngaColor}]${children}[/color]`;
    }
  }
  switch (node.tag) {
  case 'br':
    return '\n';
  case 'hr':
    return '\n======\n';
  case 'figure':
    return `\n${children}\n`;
  case 'table':
    return `\n[${node.tag}]${children}\n[/${node.tag}]\n`;
  case 'tbody':
  case 'thead':
    return children;
  case 'tr':
    return `\n[${node.tag}]${children}\n[/${node.tag}]`;
  case 'th':
  case 'td': {
    const colspan = getNumAttr(node, 'colspan');
    const colspanStr = colspan ? ` colspan=${colspan}` : '';
    const rowspan = getNumAttr(node, 'rowspan');
    const rowspanStr = rowspan ? ` rowspan=${rowspan}` : '';
    // const boldChildren = node.tag === 'th' ? `[b]${children}[/b]` : children;

    return `\n[td${rowspanStr}${colspanStr}]${children}[/td]`;
  }
  case 'h1':
  case 'h2':
  case 'h3':
  case 'h4':
  case 'h5': {
    const sizes = {
      h1: 150,
      h2: 140,
      h3: 130,
      h4: 120,
      h5: 110,
    };

    return children ? `\n[size=${sizes[node.tag]}%][b]${children}[/b][/size]\n` : '';
  }

  case 'picture':
  case 'span':
  case 'a':
  case 'div':
  case 'font':
    return children;

  case 'p':
    return children ? ensureNewlines(children, 1) : '';
  case 'strong':
  case 'b':
    return children ? `[b]${children}[/b]` : '';
  case 's':
    return children ? `[del]${children}[/del]` : '';
  case 'u':
    return children ? `[u]${children}[/u]` : '';
  case 'em':
  case 'i':
    return `[i]${children}[/i]`;
  case 'sup':
    return `[sup]${children}[/sup]`;
  case 'sub':
    return `[sub]${children}[/sub]`;
  case 'blockquote':
    return `\n[quote]${ensureNewlines(children, 1, 1)}[/quote]`;
  case 'article':
    return children;
  case 'ul': {
    const prefix = context.$prev ? '' : '\n';

    return `\n${prefix}[list]${ensureNewlines(children, 0, 1)}[/list]\n`;
  }
  case 'ol': {
    const prefix = context.$prev ? '' : '\n';

    return `\n${prefix}[list=1]${ensureNewlines(children, 0, 1)}[/list]\n`;
  }
  case 'li': {
    const prefix = context.$prev ? '\n' : '';

    return `${prefix}[*]${children}`;
  }

  case 'img': {
    const src = findAttr(node, 'src');
    if (src) {
      return `\n[img]${src.value}[/img]\n`;
    }

    return '';
  }

  default:
    console.warn(`Unhandled node type ${node.tag}`);

    return children;
  }
}

function generateBbsNodeList(nodes: Node[]) {
  let result = '';
  let prevNode: Node;
  let prevNewlines = 0;
  _.forEach(nodes, (node: Node) => {
    const nodeText = generateBbsCode(node, { $prev: prevNode });
    if (nodeText.length !== 0) {
      prevNode = node;
      const [trimmedNodeText, leftNewlines, rightNewlines] = trimNewlines(nodeText);
      // console.log('trimmed', prevNewlines, leftNewlines, rightNewlines);
      result += _.repeat('\n', Math.max(prevNewlines, leftNewlines));
      result += trimmedNodeText;
      prevNewlines = rightNewlines;
    } else if (prevNode && prevNode.kind !== 'text') {
      result += ' ';
    }
  });

  return (result + _.repeat('\n', prevNewlines))
    .replace(/ *\n/g, '\n');
}
