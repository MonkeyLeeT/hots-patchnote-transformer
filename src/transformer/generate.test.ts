import * as parse5 from 'parse5';

import { generateBbsCode, GenerationChild, genTreeToString } from './generate';
import { standardizeTree } from './html';

describe('test genTreeToString', () => {
  test('Test when tag inner NL is longer than children outer NL', () => {
    const node: GenerationChild = {
      tag: 'a',
      newlines: { inStart: 2, inEnd: 2 },
      keepEmpty: true,
      children: [
        {
          tag: 'b',
          newlines: { outBefore: 1, outAfter: 1 },
          keepEmpty: true,
        },
        {
          tag: 'c',
          newlines: { outBefore: 2, outAfter: 1 },
          keepEmpty: true,
        },
      ],
    };
    expect(genTreeToString(node)).toEqual(
      [
        '[a]\n\n[b][/b]\n\n[c][/c]\n\n[/a]',
        0,
        0,
      ],
    );
  });

  test('Test when tag inner NL is shorter than children outer NL', () => {
    const node: GenerationChild = {
      tag: 'a',
      newlines: { inStart: 1, inEnd: 1 },
      keepEmpty: true,
      children: [
        {
          tag: 'b',
          newlines: { outBefore: 2, outAfter: 2 },
          keepEmpty: true,
        },
        {
          tag: 'c',
          newlines: { outBefore: 1, outAfter: 2 },
          keepEmpty: true,
        },
      ],
    };
    expect(genTreeToString(node)).toEqual(
      [
        '[a]\n\n[b][/b]\n\n[c][/c]\n\n[/a]',
        0,
        0,
      ],
    );
  });

  test('Purge empty children except when keepEmpty is true', () => {
    const node: GenerationChild = {
      tag: 'a',
      children: [
        {
          tag: 'b',
          children: [],
        },
        {
          tag: 'c',
          keepEmpty: true,
          children: ['   '],
        },
      ],
    };
    expect(genTreeToString(node)).toEqual(
      [
        '[a][c]   [/c][/a]',
        0,
        0,
      ],
    );
  });

  test('Render <br /> using outSum', () => {
    const node: GenerationChild = {
      tag: 'a',
      children: [
        {
          children: ['123'],
        },
        {
          keepEmpty: true,
          newlines: { outSum: 1 },
        },
        {
          children: ['456'],
        },
      ],
    };
    expect(genTreeToString(node)).toEqual(
      [
        '[a]123\n456[/a]',
        0,
        0,
      ],
    );
  });

});

describe('test generateBbsCode()', () => {
  test('correctly translate menu', () => {
    const htmlNode = <parse5.DefaultTreeDocumentFragment>parse5.parseFragment(
      `<ul>
	<li><span style="font-size: 14px;"><a href="#heroes">Heroes</a></span></li>
	<li><span style="font-size: 14px;"><a href="#bug-fixes">Bug Fixes</a></span></li>
</ul>
`,
    );
    const node = standardizeTree(htmlNode.childNodes[0]);
    expect(generateBbsCode(node)).toMatchSnapshot();
  });

});
