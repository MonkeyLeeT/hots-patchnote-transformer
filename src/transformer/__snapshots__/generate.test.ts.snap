// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test generateBbsCode() correctly translate empty list to empty string 1`] = `""`;

exports[`test generateBbsCode() correctly translate menu in desired format 1`] = `
"[list][*]Heroes
[*]Bug Fixes
[/list]"
`;
